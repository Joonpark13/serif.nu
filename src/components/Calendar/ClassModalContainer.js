import { connect } from 'react-redux';
import toJS from 'util/to-js';
import { removeSection } from 'actions';
import ClassModal from './ClassModal';

/* istanbul ignore next */
function mapDispatchToProps(dispatch) {
  return {
    removeSection: (sectionId, sectionColor) => dispatch(removeSection(sectionId, sectionColor)),
  };
}

const ClassModalContainer = connect(null, mapDispatchToProps)(toJS(ClassModal));

export default ClassModalContainer;
