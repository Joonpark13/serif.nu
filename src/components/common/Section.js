import React from 'react';
import PropTypes from 'prop-types';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';

function Section({ classes, leftHeaderContent, rightHeaderContent, sectionName, onClick }) {
  return (
    <Paper elevation={2} className={classes.paper} onClick={onClick}>
      <div className={classes.container}>
        <div className={classes.header}>
          <Typography variant="caption" className={classes.text}>
            {leftHeaderContent}
          </Typography>

          <Typography variant="caption" className={classes.text}>
            {rightHeaderContent}
          </Typography>
        </div>

        <Typography variant="subtitle2" className={`${classes.text} ${classes.name}`} noWrap>
          {sectionName}
        </Typography>
      </div>
    </Paper>
  );
}

Section.propTypes = {
  classes: PropTypes.shape({
    paper: PropTypes.string,
    container: PropTypes.string,
    header: PropTypes.string,
    text: PropTypes.string,
    name: PropTypes.string,
  }).isRequired,
  leftHeaderContent: PropTypes.string,
  rightHeaderContent: PropTypes.string,
  sectionName: PropTypes.string,
  onClick: PropTypes.func,
};

Section.defaultProps = {
  leftHeaderContent: '',
  rightHeaderContent: '',
  sectionName: '',
  onClick: () => {},
};

export default Section;
