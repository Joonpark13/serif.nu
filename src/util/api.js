/* eslint-disable import/prefer-default-export */
export function searchURL(query) {
  return `${process.env.API_URL}/search?term_id=4720&search_query=${query}`;
}

export function sectionsURL(termId, schoolAbbv, subjectAbbv, courseAbbv) {
  return `${process.env.API_URL}/sections?term_id=${termId}&school_abbv=${schoolAbbv}&subject_abbv=${subjectAbbv}&course_abbv=${courseAbbv}`;
}
