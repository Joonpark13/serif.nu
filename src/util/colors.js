export const northwesternPurple = '#4E2A84';
export const northwesternBrightOrange = '#FFC520';
export const northwesternRichBlack10 = '#D8D6D6';
