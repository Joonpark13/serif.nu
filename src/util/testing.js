/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import * as useSelector from './use-selector';

jest.mock('./use-selector');

// Helper function that returns a function which creates enzyme wrappers with default props
export function wrapperCreator(Component, defaultProps = {}) {
  let classes;
  return props => shallow(<Component classes={classes} {...defaultProps} {...props} />);
}

/* istanbul ignore next */
export function mockUseSelector(...args) {
  useSelector.default.mockReset();
  args.forEach(arg => useSelector.default.mockReturnValueOnce(arg));

  return useSelector.default;
}

/* istanbul ignore next */
export function mockUseDispatch() {
  const dispatchSpy = jest.fn();
  reactRedux.useDispatch.mockReturnValue(dispatchSpy);
  return dispatchSpy;
}
