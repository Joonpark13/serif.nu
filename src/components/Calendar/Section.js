import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/styles';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import { getDurationInHours, getFormattedClassEvent } from 'util/time';

export const MAX_WIDTH_PERCENT = 97;

export const styles = {
  paper: {
    position: 'absolute',
    top: ({ section }) => {
      const minute = section.event.start.minute;
      const offset = Math.round(minute / 60 * 100);
      return `${offset}%`;
    },
    left: ({ section }) => `${MAX_WIDTH_PERCENT * section.columnWidth * section.column}%`,
    height: ({ section }) => {
      const durationInHours = getDurationInHours(section.event);
      const heightInPercent = Math.round(durationInHours * 100);
      return `${heightInPercent}%`;
    },
    width: ({ section }) => `${MAX_WIDTH_PERCENT * section.columnWidth}%`,
    backgroundColor: ({ section }) => section.color,
    overflow: 'hidden',
  },
  container: {
    margin: '3px',
  },
  text: {
    color: 'white',
  },
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    flexWrap: 'wrap',
  },
};

function Section({ classes, section }) {
  return (
    <Paper className={classes.paper}>
      <div className={classes.container}>
        <div className={classes.header}>
          <Typography variant="caption" className={classes.text}>
            {getFormattedClassEvent(section.event)}
          </Typography>

          <Typography variant="caption" className={classes.text}>
            {`${section.subjectId} ${section.courseId}`}
          </Typography>
        </div>

        <Typography variant="subtitle2" className={`${classes.text} ${classes.name}`} noWrap>
          {section.name}
        </Typography>
      </div>
    </Paper>
  );
}

Section.propTypes = {
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};

export { Section as UnstyledSection };
export default withStyles(styles)(Section);
