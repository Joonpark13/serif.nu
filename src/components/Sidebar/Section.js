import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';
import { getFormattedClassSchedule } from 'util/time';

export const styles = {
  sectionTitle: {
    fontWeight: 'bold',
  },
};

function Section({ addSection, section, disabled, classes }) {
  return (
    <ListItem
      key={section.id}
      button
      onClick={() => addSection(section)}
      disabled={disabled}
    >
      <ListItemText>
        <Typography variant="h6" className={classes.sectionTitle}>
          {`Section ${section.sectionNumber}`}
        </Typography>

        {section.topic && <Typography>{section.topic}</Typography>}

        {section.schedule.map((scheduleObj, index) => (
          /* eslint-disable react/no-array-index-key */
          <Typography key={index}>
            {getFormattedClassSchedule(scheduleObj)}
          </Typography>
          /* eslint-enable react/no-array-index-key */
        ))}

        <Typography>{section.schedule[0].location}</Typography>

        <Typography>
          {section.instructors.map((teacher, idx) => (
            `${teacher}${idx !== section.instructors.length - 1 ? ', ' : ''}`))
          }
        </Typography>
      </ListItemText>
    </ListItem>
  );
}

Section.propTypes = {
  addSection: PropTypes.func.isRequired,
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  disabled: PropTypes.bool,
};

Section.defaultProps = {
  disabled: false,
};

export { Section as UnstyledSection };
export default withStyles(styles)(Section);
