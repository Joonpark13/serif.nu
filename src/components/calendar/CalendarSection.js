import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/styles';
import { getFormattedEventTime, getDurationInHours } from 'util/time';
import Section from 'components/common/Section';
import ClassModalContainer from './ClassModalContainer';

export const MAX_WIDTH_PERCENT = 97;

export const styles = {
  paper: {
    position: 'absolute',
    top: ({ section }) => {
      const minute = section.event.start.minute;
      const offset = Math.round(minute / 60 * 100);
      return `${offset}%`;
    },
    left: ({ section }) => `${MAX_WIDTH_PERCENT * section.columnWidth * section.column}%`,
    height: ({ section }) => {
      const durationInHours = getDurationInHours(section.event);
      const heightInPercent = Math.round(durationInHours * 100);
      return `${heightInPercent}%`;
    },
    width: ({ section }) => `${MAX_WIDTH_PERCENT * section.columnWidth}%`,
    backgroundColor: ({ section }) => section.color,
    overflow: 'hidden',
    cursor: ({ isPreview }) => isPreview ? 'default' : 'pointer',
    zIndex: ({ isPreview }) => isPreview ? 2 : 1,
  },
  container: {
    margin: '3px',
  },
  text: {
    color: 'white',
  },
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    flexWrap: 'wrap',
  },
};

class CalendarSection extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showDialog: false,
    };
    this.handleClick = this.handleClick.bind(this);
    this.toggleDialog = this.toggleDialog.bind(this);
  }

  handleClick() {
    const { isPreview } = this.props;
    if (!isPreview) {
      this.toggleDialog();
    }
  }

  toggleDialog() {
    this.setState(state => ({ showDialog: !state.showDialog }));
  }

  render() {
    const { classes, section } = this.props;
    const { showDialog } = this.state;

    const leftHeaderContent = getFormattedEventTime(section.event);
    const rightHeaderContent = `${section.subjectId} ${section.courseId}`;

    return (
      <div>
        <Section
          onClick={this.handleClick}
          classes={classes}
          leftHeaderContent={leftHeaderContent}
          rightHeaderContent={rightHeaderContent}
          sectionName={section.name}
        />
        <ClassModalContainer
          section={section}
          showDialog={showDialog}
          toggleDialog={this.toggleDialog}
        />
      </div>
    );
  }
}

CalendarSection.propTypes = {
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  isPreview: PropTypes.bool,
};

CalendarSection.defaultProps = {
  isPreview: false,
};

export { CalendarSection as UnstyledCalendarSection };
export default withStyles(styles)(CalendarSection);
