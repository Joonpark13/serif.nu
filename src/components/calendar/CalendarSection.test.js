import React from 'react';
import { shallow } from 'enzyme';
import { mockStyles } from 'util/testing';
import Section from 'components/common/Section';
import ClassModalContainer from './ClassModalContainer';
import { UnstyledCalendarSection, styles, MAX_WIDTH_PERCENT } from './CalendarSection';

describe('CalendarSection', () => {
  const dow = 'Mo';
  const event = {
    dow,
    start: {
      hour: 10,
      minute: 30,
    },
    end: {
      hour: 12,
      minute: 0,
    },
  };

  describe('dynamic styles', () => {
    const hour = 10;
    const color = 'some color';
    const section = { event, color, column: 0, columnWidth: 1 };

    it('correctly calculates section card placement', () => {
      expect(styles.paper.top({ hour, dow, section })).toBe('50%');
    });

    it('correctly calculates section card height', () => {
      expect(styles.paper.height({ hour, dow, section })).toBe('150%');
    });

    it('correctly grabs the section background color', () => {
      expect(styles.paper.backgroundColor({ section })).toBe(color);
    });

    it('correctly calculates left offset', () => {
      expect(styles.paper.left({ section })).toBe('0%');
    });

    it('correctly calculates width', () => {
      expect(styles.paper.width({ section })).toBe(`${MAX_WIDTH_PERCENT}%`);
    });

    it('correctly sets zIndex to 1 when class is not preview', () => {
      expect(styles.paper.zIndex({ isPreview: false })).toBe(1);
    });

    it('correctly sets zIndex to 2 when class is preview', () => {
      expect(styles.paper.zIndex({ isPreview: true })).toBe(2);
    });

    it('correctly sets cursor to pointer when class is not preview', () => {
      expect(styles.paper.cursor({ isPreview: false })).toBe('pointer');
    });

    it('correctly sets cursor to default when class is preview', () => {
      expect(styles.paper.cursor({ isPreview: true })).toBe('default');
    });
  });

  it('renders correctly', () => {
    const classes = mockStyles(styles);
    const testSection = { id: '12345', course: '101-1', event };
    const wrapper = shallow(
      <UnstyledCalendarSection section={testSection} classes={classes} />,
    );

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('renders modal correctly', () => {
    const classes = mockStyles(styles);
    const testSection = { id: '12345', course: '101-1', event };
    const wrapper = shallow(
      <UnstyledCalendarSection section={testSection} classes={classes} />,
    );

    wrapper.find(Section).simulate('click');

    expect(wrapper.find(ClassModalContainer).prop('showDialog')).toBe(true);
  });

  it('does nothing as a preview section when clicked', () => {
    const classes = mockStyles(styles);
    const testSection = { id: '12345', course: '101-1', event };
    const wrapper = shallow(
      <UnstyledCalendarSection section={testSection} classes={classes} isPreview />,
    );

    wrapper.find(Section).simulate('click');

    expect(wrapper.find(ClassModalContainer).prop('showDialog')).toBe(false);
  });
});
