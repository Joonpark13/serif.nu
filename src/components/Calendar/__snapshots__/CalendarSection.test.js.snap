// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarSection renders correctly 1`] = `
<div>
  <Section
    classes={
      Object {
        "container": "",
        "header": "",
        "paper": "",
        "text": "",
      }
    }
    leftHeaderContent="10:30 - 12:00"
    onClick={[Function]}
    rightHeaderContent="undefined undefined"
    sectionName=""
  />
  <Connect(Component)
    section={
      Object {
        "course": "101-1",
        "event": Object {
          "dow": "Mo",
          "end": Object {
            "hour": 12,
            "minute": 0,
          },
          "start": Object {
            "hour": 10,
            "minute": 30,
          },
        },
        "id": "12345",
      }
    }
    showDialog={false}
    toggleDialog={[Function]}
  />
</div>
`;
